src/signatures/secp256k1.rs

```diff
@@ -0,0 +1,168 @@
+pub type FieldBytes = k256::FieldBytes;
+
+/// Secp256k1 Shared Secret - the result of a Diffie-Hellman key exchange.
+///
+/// Each party computes this from an (ephemeral) [SecretKey] and their counterparty's [PublicKey].
+pub type SharedSecret = k256::ecdh::SharedSecret;
+
+/// Secp256k1 Public Key
+pub struct PublicKey(k256::PublicKey);
+
+// algorithm identifier used for errors
+const NAME: &str = "Secp256k1";
+
+impl PublicKey {
+    /// Load a [PublicKey] from a slice of bytes.
+    pub fn from_bytes(bytes: &[u8]) -> crate::Result<Self> {
+        assert_buffer_eq!(bytes.len(), UntaggedPointSize::<Secp256k1>::USIZE, "bytes");
+
+        EncodedPoint::from_untagged_bytes(bytes.into())
+            .decode()
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "bytes",
+                to: "Secp256k1 Public Key",
+            })
+    }
+
+    /// Load a [PublicKey] from big-endian serialized coordinates.
+    pub fn from_coord(x: &[u8], y: &[u8]) -> crate::Result<Self> {
+        let x: &FieldBytes = x.try_into().map_err(|_| crate::Error::ConvertError {
+            from: "bytes",
+            to: "Secp256k1 Point (x)",
+        })?;
+
+        let y: &FieldBytes = y.try_into().map_err(|_| crate::Error::ConvertError {
+            from: "bytes",
+            to: "Secp256k1 Point (y)",
+        })?;
+
+        EncodedPoint::from_affine_coordinates(x, y, false)
+            .try_into()
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "Coordinates",
+                to: "Secp256k1 Public Key",
+            })
+    }
+
+    /// Return the [PublicKey] as an array of bytes.
+    pub fn to_bytes(&self) -> GenericArray<u8, UntaggedPointSize<Secp256k1>> {
+        let point: EncodedPoint<Secp256k1> = self.0.to_encoded_point(false);
+
+        // sanity check encoded point
+        assert!(!point.is_compressed());
+        assert!(!point.is_identity());
+
+        // unwrap is okay - non-identity points are guaranteed by `PublicKey`.
+        point.to_untagged_bytes().unwrap()
+    }
+
+    /// Serialize this public key as SEC1 encoded coordinates.
+    pub fn to_coord(&self) -> crate::Result<(FieldBytes, FieldBytes)> {
+        match self.0.to_encoded_point(false).coordinates() {
+            Coordinates::Uncompressed { x, y } => Ok((*x, *y)),
+            Coordinates::Identity | Coordinates::Compressed { .. } => Err(crate::Error::ConvertError {
+                from: "Secp256k1 Public Key",
+                to: "Coordinates",
+            }),
+        }
+    }
+
+    /// Verify that the given `message` and `signature` are authentic.
+    pub fn verify(&self, message: &[u8], signature: &[u8]) -> crate::Result<()> {
+        signature::Signature::from_bytes(signature)
+            .and_then(|signature| Verifier::verify(self, message, &signature))
+            .map_err(|_| crate::Error::SignatureError { alg: NAME })
```

Comment:
Authentication, verification, integrity check errors should stand out / be more alarming because they are caused by external causes, adversaries.